Bounded loops

b-mueller opened the pull request:
Adds a new search strategy called "bfs-bounded". When an address is repeatedly targeted by a JUMPI in path (`>JUMPDEST_LIMIT` times which is currently set to 4), this strategy ignores subsequent JUMPI instructions that target the same destination.

Solves #1048 for the most part (it's still quite slow on byte array copy because memory operations are slow, but with the number of iteration reduced the time Mythril spends on this should become reasonable). Should also help with BECToken and other samples that contain loops.

norhh commented on mythril/analysis/modules/dos.py:
```diff
@@ -70,7 +70,7 @@ def _analyze_states(self, state: GlobalState) -> List[Issue]:
                 try:
                     self._jumpdest_count[transaction][target] += 1
-                    if self._jumpdest_count[transaction][target] == 4:
+                    if self._jumpdest_count[transaction][target] == 3:
```
Using JUMPDEST_LIMIT - 1 might be good here, what do you think?

b-mueller replied:
Yeah, great idea!

norhh commented on mythril/laser/ethereum/strategy/custom.py:
```diff
@@ -0,0 +1,55 @@
+
+        opcode = state.get_current_instruction()["opcode"]
+
+        if opcode == "JUMPI":
```
We can reduce the nesting here by returning if the `opcode != "JUMPI"`, less nesting looks cool and easy to read.

norhh commented:
This can actually be placed in a plugin right?

b-mueller commented:
I guess? Joran's suggestion was to put it into a strategy.

norhh commented:
Maintaining a global counter for all the states may not be a good idea right? Like if there are 3 states in worklist, all hit the loop, then self._jumpcount[tx][address] ++ happens globally right? are the transaction stacks deepcopied each time?

b-mueller commented:
Yes, the strategy counts jumpdests globally (per transaction) for the whole symbolic execution run, `_jumpcount` is a member of the `BFSBoundedLoopsStrategy` class instance, how is that related to transaction stacks? Sorry if I'm not following correctly :)

norhh commented:
What if there are 10 states in worklist which enter the same JUMPI condition(which is an if condition rather than a loop), In that case there might be a problem right?

b-mueller commented:
In what cases would that happen? The only case I can think of is complex if-statements if they compile in certain ways? 

```
if (a == 0 | a == 1 | a == 2 | a == 3 | a == 4) {

} else {

}
```

Would be interesting to have some concrete examples where it causes issues.

norhh commented:
```
...
if()
{}
else {}
// there will be 2 states here

for(i=0; i<num; i++){} 
// there will be 2*JUMPI_COUNT states here

if(condition) //some states above might get pruned here.
{} 
 ...
```
It might be uncommon but still might be possible :cry: .(uncommon but likely to be encountered given that we analyze lot's of contracts per day)

b-mueller commented:
Ok, I confirmed that this is exactly the case. With this strategy we can detect this:

```
pragma solidity ^0.5.0;


contract Lol {

	function lol(uint256 a) public {

		if ((a == 0) || (a == 1) || (a == 2) || (a == 3)) {
		   if (a == 3) {
		      selfdestruct(msg.sender);
		  	}
		 }
	} 
}
```

But not this:

```
pragma solidity ^0.5.0;

contract Lol {

	function lol(uint256 a) public {

		if ((a == 0) || (a == 1) || (a == 2) || (a == 3) || (a == 4)) {
		   if (a == 4) {
		      selfdestruct(msg.sender);
		  	}
		 }
	} 
}
```

Although, in this case, you could say that it is very similar to a loop iterating over `a`. I guess with this strategy we'll miss cases whenever the code searches for a specific value in a variable or some complex combination of conditions.

The most straightforward way to deal with this IMO is to select a proper bound (perhaps a bit higher than 4?) and accept that we can miss some paths of there's an unbounded amount of values, in which case fuzzing would be more appropriate (and this is something we could hand over to Harvey)?

JoranHonig commented:
I'd add a state annotation which tracks which instructions have been executed over a given path.
Additionally, I'd say this should be an extending strategy (similar to how the coverage strategy is implemented). 
That way we don't have to implement the same code all the time

JoranHonig commented on mythril/analysis/modules/dos.py:
```diff
@@ -4,6 +4,7 @@
 from typing import Dict, cast, List
 
 from mythril.analysis.swc_data import DOS_WITH_BLOCK_GAS_LIMIT
+from mythril.laser.ethereum.strategy.custom import JUMPDEST_LIMIT
```
Analysis code should not be dependent on search strategies, keeping them decoupled makes maintenance & further development easier.

norhh replied:
Then let's have a constants file for the laser module.That might be cool.

JoranHonig replied:
There doesn't need to be a dependency here

norhh replied:
The JUMPDEST_LIMIT is dependent in both areas right? so this should have a dependece with laser.

b-mueller replied:
Yeah, the idea here is to make sure that the limit doesn't matches the loop bound because otherwise the it will never be triggered. But we can remove the dependency.

b-mueller replied:
Actually, now that the search strategy sets a `JumpdestCountAnnotation` this module could be simplified further by accessing that annotation instead of maintaining its own counter. However this would mean that the dependence remains.

b-mueller replied:
No, that's actually BS, because a different search strategy could be used.

JoranHonig commented:
@norhh I think a search strategy is perfect for this use case

I'd use a plugin if we extend the behaviour of laser in some way ( for example with the coverage ).
A search strategy interface is limited, making it easier to implement and more simple ( I hope )

norhh commented:
something like this might fail
```


contract Lame {

    uint bata;

    function addCreditors(uint a) public returns (bool) {
        if(a == 0)
        {
            bata += 1;
        }
        else {
            bata -= 1;
        }
        // there will be 2 states here
        uint i;
        for(i=0; i<2; i++){
            a+=1;
            bata += a;
        } 

        if(a == 3) // some states might get pruned here
        {
            selfdestruct(msg.sender);
        } 
    }

}

```
`I'd add a state annotation which tracks which instructions have been executed over a given path`
That should do the work.

norhh commented:
@JoranHonig .I was saying something like the coverage strategy plugin ,We can either place it as a plugin or an extensible class, as long as we can have a cool flag for it.

b-mueller commented:
> something like this might fail
>   contract Lame {

Yeah, basically another variant where we'd get state explosion because `a`\`bata` are incremented in a loop and each value of a could potentially trigger a different path. My intuition is that Harvey could deal with this better? @wuestholz

JoranHonig commented:
symbolic execution shouldn't have a problem with this contract

wuestholz commented:
@b-mueller Yeah, Harvey deals with path/state explosion a bit more gracefully. The main reason is that it explores the different paths and states less systematically. This avoids getting stuck in one part of the state space by exploring only its surrounding states.

b-mueller commented:
> I'd add a state annotation which tracks which instructions have been executed over a given path

That sounds reasonable. Not sure if refactoring this to a plugin is better, I'll finish up the class fist :)

JoranHonig commented:
The difference between a plugin and a search strategy is that a plugin has an initialization phase where it is allowed to introduce hooks in the laser vm.
A search strategy doesn't introduce hooks, but is initialized with the work list & called each time the executor wants to know a state to work on.

I don't think a plugin is necessary for this case. The coverage based search strategy is based in a plugin because it is dependent on that plugin. I'm not sure whether this is the best way to go about indicating the dependency, but this is what I decided to add at that point.

b-mueller commented:
I rewrote this to use a StateAnnotation. Not sure how extending strategies are meant to be used / invoked?

norhh commented on mythril/analysis/modules/dos.py:
```diff
@@ -70,7 +71,7 @@ def _analyze_states(self, state: GlobalState) -> List[Issue]:
                 try:
                     self._jumpdest_count[transaction][target] += 1
-                    if self._jumpdest_count[transaction][target] == 4:
+                    if self._jumpdest_count[transaction][target] == JUMPDEST_LIMIT - 1:
```
this would always go in for if conditions right? if JUMPDEST_LIMIT is 2

b-mueller replied:
no always but if there's more than 1 condition. So probably 4 is a more reasonable limit...

b-mueller commented:
@JoranHonig I don't see `CoverageStrategy` used anywhere in the code. How are extended strategies supposed to be selected? Do we still need to add CLI support for those?

b-mueller commented:
I refactored this into an extension similar to the coverage strategy and added a method to laser for loading it "on top of" its existing strategy. Also added a `--loop-bound` CLI argument with a default value of 4 which is passed through to SymExecWrapper.

Should be ready for review now...

norhh commented on mythril/laser/ethereum/strategy/extensions/bounded_loops.py:
```diff
@@ -0,0 +1,81 @@
+            )
+        )
+
+        BasicSearchStrategy.__init__(
```
We can use `super()` here. It would be more readable as super means parent class

b-mueller replied:
super() does not work here because of the way search strategies are passed to laser and initialized. See also [CoverageStrategy](https://github.com/ConsenSys/mythril/blob/develop/mythril/laser/ethereum/plugins/implementations/coverage/coverage_strategy.py).

norhh commented on mythril/laser/ethereum/strategy/extensions/bounded_loops.py:
```diff
@@ -0,0 +1,81 @@
+        :return:
+        """
+
+        while 1:
```
```suggestion
        while True:
```

norhh commented on mythril/laser/ethereum/strategy/extensions/bounded_loops.py:
```diff
@@ -0,0 +1,81 @@
+            else:
+                annotation = annotations[0]
+
+            target = int(util.get_concrete_int(state.mstate.stack[-1]))
```
wouldn't there be a chance of a ValueError? sometimes jumpi can be arbitrary.

b-mueller replied:
I think it's TypeError right?

norhh replied:
Oh! yeah.

norhh commented on mythril/interfaces/cli.py:
```diff
@@ -197,13 +197,20 @@ def create_parser(parser: argparse.ArgumentParser) -> None:
+        type=int,
+        default=4,
+        help="Bound loops at n iterations",
+        metavar="N",
```
Should we have metavar as N here? as that would be used for display of the arg.We can just skip it.

b-mueller replied:
I thought it would be cool to say "n iterations" and call the metavar N 😀 Can we leave it like that?

norhh commented on mythril/analysis/modules/dos.py:
```diff
@@ -70,7 +70,7 @@ def _analyze_states(self, state: GlobalState) -> List[Issue]:
                 try:
                     self._jumpdest_count[transaction][target] += 1
-                    if self._jumpdest_count[transaction][target] == 4:
+                    if self._jumpdest_count[transaction][target] == 3:
```
We should use that constant here right? Propogating that to here might be a problem.

b-mueller replied:
Unfortunately there's currently no way to pass arguments to modules or access svm variables from it. I'd leave this as a hardcoded value for now. Later, we could have Mythril set a ModuleConfigAnnotation as a new feature?
